### ISSUE TITLE
duplicate devices

### ISSUE THREAD

**Issue by stintel** [matrix-org/dendrite#2693]
### Background information

- **Dendrite version or git SHA**: 0.9.6
- **Monolith or Polylith?**: monolith
- **SQLite3 or Postgres?**: postgres
- **Running in Docker?**: no
- **`go version`**: 1.18.5
- **Client used (if applicable)**: Neochat 1.2.0

### Description
 - **What** is the problem: duplicate devices for same client
 - **Who** is affected: neochat clients on my server
 - **How** is this bug manifesting: when using Neochat, a lot of duplicate devices are appearing, sometimes every 5 minutes
 - **When** did this first appear: for as long as I can remember, started using dendrite at version 0.3.4 in December 2020, neochat at version 1.0.1 in February 2021

### Steps to reproduce

- install dendrite
- connect a neochat client
- wait for duplicate devices to appear

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from server logs, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.

Alternatively, please send logs to @kegan:matrix.org or @neilalexander:matrix.org
with a link to the respective Github issue, thanks!
-->
While possibly a neochat bug, I belive Dendrite should have a configurable limit for max # of devices per user to avoid situations like this. The default should be reasonably low. I have just signed out >600 (!) devices, both dendrite and postgresql where eating all 4 vCPUs after doing so, with load average growing over 10.

I believe this could be related to my Dendrite instance permanently eating a CPU core and consuming consistently more than 1GB of RAM, and random rooms being completely out of sync for the last few months. Possibly also the cause of #2566? I have had that problem many times, and deleting the jetstream directory seemed to bring temporary relief, but eventually some channels stopped syncing again.

![image](https://user-images.githubusercontent.com/1258811/188335338-50772303-78c9-40c1-8c14-65a7fab99de4.png)

**Comment by stintel** [comment]
Not so long after opening this issue the dendrite-monolith-server process is back to eating a CPU core and > 1GB RAM, but at least clients seem to respond normally now, seeing proper room names and no longer crazy slow.


I wanted to ask this in the Dendrite room, but it appears my server is banned from the room (already in contact with support@ about that), so I'll ask it here:
Do the profile files generated from the instructions on https://matrix-org.github.io/dendrite/development/profiling contain sensitive data?

**Comment by neilalexander** [comment]
The profiles are safe, so might be worth inspecting what's going on there.

**Comment by stintel** [comment]
> Has this bug been reported to the NeoChat team? Something is horribly wrong if it's registering new devices every few minutes.

Actually the client causing this might not have been NeoChat. It appears to be matrix-ircd, which for some reason shows up as a different client. After enabling real_ip_header I now see the correct client IP, which confirms it's matrix-ircd and not NeoChat.

**Comment by kegsay** [comment]
> I believe the possibility to limit the amount of devices a user can register is a reasonable feature request?

YES PLEASE. The problem is that if you omit the `device_id` field on registration/login, one is randomly generated. So just innocently missing this field can cause rampant device creation. This then exacerbates slow E2EE because there's so many devices to encrypt for.
